pkg/prowloader/prow.go

```diff
@@ -623,30 +711,44 @@ func (pl *ProwLoader) findOrAddPullRequests(refs *prow.Refs, pjPath string) []mo
 	return pulls
 }
 
-func (pl *ProwLoader) findOrAddTest(name string) uint {
+func (pl *ProwLoader) findOrAddTest(name string) (uint, error) {
+	pl.prowJobRunTestCacheLock.RLock()
 	if id, ok := pl.prowJobRunTestCache[name]; ok {
-		return id
+		pl.prowJobRunTestCacheLock.RUnlock()
+		return id, nil
 	}
+	pl.prowJobRunTestCacheLock.RUnlock()
 
-	test := &models.Test{}
-	pl.dbc.DB.Where("name = ?", name).Find(&test)
-	if test.ID == 0 {
-		test.Name = name
-		pl.dbc.DB.Save(test)
+	// Lock the whole block here because composite operations (like FirstOrCreate) in gorm are
+	// NOT atomic. https://github.com/go-gorm/gorm/issues/6253
+	pl.prowJobRunTestCacheLock.Lock()
+	defer pl.prowJobRunTestCacheLock.Unlock()
+	test := &models.Test{
+		Name: name,
+	}
+	tx := pl.dbc.DB.FirstOrCreate(&test)
+	if tx.Error != nil {
+		return 0, tx.Error
 	}
+
 	pl.prowJobRunTestCache[name] = test.ID
-	return test.ID
+	return test.ID, nil
 }
 
 func (pl *ProwLoader) findSuite(name string) *uint {
 	if name == "" {
 		return nil
 	}
 
+	pl.suiteCacheLock.RLock()
 	if id, ok := pl.suiteCache[name]; ok {
+		pl.suiteCacheLock.RUnlock()
 		return id
 	}
+	pl.suiteCacheLock.RUnlock()
 
+	pl.suiteCacheLock.Lock()
+	defer pl.suiteCacheLock.Unlock()
```

Comment:
Looks like an unnecessary lock/unlock here.

Reply:
It's needed since we're manipulating the cache a little bit further down